File: src/fmj/goods/GoodsMedicineChg4Ever.kt
package fmj.goods

import fmj.characters.Player

/**
 * 11仙药类
 * 永久性改变人物属性
 * @author Chen
 */
class GoodsMedicineChg4Ever : BaseGoods(), IEatMedicine {

    private var mMpMax: Int = 0
    private var mHpMax: Int = 0
    private var mdf: Int = 0
    private var mat: Int = 0
    private var mling: Int = 0
    private var mSpeed: Int = 0
    private var mLuck: Int = 0

    override fun setOtherData(buf: ByteArray, offset: Int) {
        mMpMax = get1ByteSInt(buf, offset + 0x16)
        mHpMax = get1ByteSInt(buf, offset + 0x17)
        mdf = get1ByteSInt(buf, offset + 0x18)
        mat = get1ByteSInt(buf, offset + 0x19)
        mling = get1ByteSInt(buf, offset + 0x1a)
        mSpeed = get1ByteSInt(buf, offset + 0x1b)
        mLuck = get1ByteSInt(buf, offset + 0x1d)
    }

    override fun eat(player: Player) {
        player.maxMP = player.maxMP + mMpMax
        player.maxHP = player.maxHP + mHpMax
        player.defend = player.defend + mdf
        player.attack = player.attack + mat
        player.lingli = player.lingli + mling
        player.speed = player.speed + mSpeed
        player.luck = player.luck + mLuck
        Player.sGoodsList.deleteGoods(type, index)
    }
}

File: packages/core/src/goods/goods-medicine-chg4ever.ts
import type { Player } from '@/characters';
import { BaseGoods, type BaseGoodsData } from './base-goods';

export interface GoodsMedicineChg4EverData extends BaseGoodsData {
  readonly mpMax: number;
  readonly hpMax: number;
  readonly defense: number;
  readonly attack: number;
  readonly spirit: number;
  readonly agility: number;
  readonly luck: number;
}

export class GoodsMedicineChg4Ever extends BaseGoods {
  mpMax: number;
  hpMax: number;
  defense: number;
  attack: number;
  spirit: number;
  agility: number;
  luck: number;

  constructor(data: GoodsMedicineChg4EverData) {
    super(data);
    this.mpMax = data.mpMax;
    this.hpMax = data.hpMax;
    this.defense = data.defense;
    this.attack = data.attack;
    this.spirit = data.spirit;
    this.agility = data.agility;
    this.luck = data.luck;
  }

  eat(player: Player): boolean {
    player.maxMp += this.mpMax;
    player.maxHp += this.hpMax;
    player.defense += this.defense;
    player.attack += this.attack;
    player.spirit += this.spirit;
    player.agility += this.agility;
    player.luck += this.luck;
    return true;
  }
}
